### PR TITLE
feat: ListItem component

### PR DESCRIPTION
**Pull request by joshhowenstine**
## Description

ListItem component for Solid

### CODE REVIEW

**Review by chiefcll — APPROVED** [review]
Nice work. Excited for the next component

File: src/packages/solid/components/ListItem/ListItem.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  tone?: Tone;
+  style?: Partial<ListItemStyles>;
+}
+
+const ListItem: Component<ListItemProps> = (props) => {
+  return (
+    <View
+      {...props}
+      style={[
+        ...[props.style].flat(),
```
[the remark this answers is not in the document]
Comment (anthony9187) [reply src/packages/solid/components/ListItem/ListItem.tsx:18]: yeah this is more of a stop-gap for now to ensure we can pass styles in. we need to come up with a pattern/rules for how and what properties we'll support in components

File: src/packages/solid/components/ListItem/ListItem.styles.ts [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+const description: ListItemConfig = {
+  themeKeys: {
+    color: 'textColor'
+  },
+  base: {
+    color: theme.color.textNeutral,
+    textAlign: 'left',
+    x: container.base.paddingX,
+    y: title.base.lineHeight + title.base.y,
```
Comment (anthony9187) [inline src/packages/solid/components/ListItem/ListItem.styles.ts:138]: i think we'll need to find another way to accomplish this - as this is written it doesn't support tones or componentConfig styles. are these themeable in LUI2? if not i think we should calculate these x and y values inside the component code

File: src/packages/solid/utils/getWidthByUpcount.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+export function getWidthByUpCount(theme, upCount = 1) {
```
Comment (anthony9187) [inline src/packages/solid/utils/getWidthByUpcount.js:1]: we need to add the license header here. this also should be written in TS or have a relevant d.ts file if that isn't possible

File: src/packages/solid/components/ListItem/ListItem.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import { type Component } from 'solid-js';
```
Comment (anthony9187) [inline src/packages/solid/components/ListItem/ListItem.tsx:1]: license header

File: src/packages/solid/components/ListItem/ListItem.styles.ts [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+import type { Tone } from '../../types/types.js';
+import type { ComponentStyleConfig, NodeStyleSet, TextStyleSet } from '../../types/types.js';
+import { makeComponentStyles } from '../../utils/index.js';
+import { getWidthByUpCount } from '../../utils/getWidthByUpcount.js';
+
+export interface ListItemStyles {
+  tone: Tone,
+  Container: NodeStyleSet;
+  Title: TextStyleSet;
+  Description: TextStyleSet;
```
Comment (anthony9187) [inline src/packages/solid/components/ListItem/ListItem.styles.ts:29]: any property we're returning to the component that isn't included inside of `NodeStyles` needs to be added to NodeStyleSet
ex.
```
Container: NodeStyleSet<{ paddingX: number }>;
```

**Comment by anthony9187** [comment]
I'm seeing a lot of linting errors, can you run it before we merge?

File: src/packages/solid/components/ListItem/ListItem.styles.ts (hunk 2 above)
Comment (ImCoolNowRight) [reply src/packages/solid/components/ListItem/ListItem.styles.ts:138]: The `x` value is themeable in LUI, but the Y is not. All of the content in ListItem is just centered vertically.

File: src/packages/solid/components/CardContent/CardContent.styles.ts [hunk 6]
```diff
@@ -54,7 +54,7 @@ const container: CardContentConfig = {
   },
   modes: {},
   tones: {},
-  themeStyles: cardContentthemeStyles
+  themeStyles: themeStyles
```
Comment (chiefcll) [inline src/packages/solid/components/CardContent/CardContent.styles.ts:57]: can shortcut this as `themeStyles,`

File: src/packages/solid/components/ListItem/ListItem.styles.ts [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    },
+    brand: {
+      color: theme.color.interactiveBrand,
+      focus: {
+        color: theme.color.fillNeutral
+      }
+    }
+  },
+  themeStyles: {
+    ...surfaceThemeStyles,
```
Comment (chiefcll) [inline src/packages/solid/components/ListItem/ListItem.styles.ts:91]: can we do this once at the top? - create a `const themeStyles = { ... }` and then just reuse it

File: src/packages/solid/components/ListItem/ListItem.styles.ts [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+    },
+    brand: {
+      color: theme.color.fillBrand,
+      focus: {
+        color: theme.color.fillBrand
+      }
+    }
+  },
+  themeStyles: {
+    ...surfaceThemeStyles,
```
Comment (chiefcll) [inline src/packages/solid/components/ListItem/ListItem.styles.ts:126]: here

File: src/packages/solid/components/ListItem/ListItem.styles.ts [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
```
Comment (chiefcll) [inline src/packages/solid/components/ListItem/ListItem.styles.ts:162]: here

File: src/packages/solid/components/ListItem/ListItem.tsx [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+  );
+};
+
+const ListItemContainer: Component<ListItemContainerProps> = props => {
+  const containerWidth = createMemo(() => getContainerWidth(props, styles));
+  return (
+    <View
+      {...props}
+      width={containerWidth()}
+      // @ts-expect-error TODO style type needs to be updated to accept nested arrays
```
Comment (chiefcll) [inline src/packages/solid/components/ListItem/ListItem.tsx:94]: this works now?

File: src/packages/solid/components/ListItem/ListItem.styles.ts (hunk 7 above)
Comment (anthony9187) [reply src/packages/solid/components/ListItem/ListItem.styles.ts:91]: good call!

File: src/packages/solid/components/ListItem/ListItem.tsx (hunk 10 above)
Comment (anthony9187) [reply src/packages/solid/components/ListItem/ListItem.tsx:94]: it works in the JS, this is just silencing the TS error

Screenshot 2024-04-23 at 9 41 25 AM

Comment (anthony9187) [reply src/packages/solid/components/ListItem/ListItem.tsx:94]: (or the text version)
```
Type '(IntrinsicNodeStyleProps | (IntrinsicNodeStyleProps | undefined)[] | undefined)[]' is not assignable to type 'IntrinsicNodeStyleProps | (IntrinsicNodeStyleProps | undefined)[] | undefined'.
  Type '(IntrinsicNodeStyleProps | (IntrinsicNodeStyleProps | undefined)[] | undefined)[]' is not assignable to type '(IntrinsicNodeStyleProps | undefined)[]'.
    Type 'IntrinsicNodeStyleProps | (IntrinsicNodeStyleProps | undefined)[] | undefined' is not assignable to type 'IntrinsicNodeStyleProps | undefined'.
      Type '(IntrinsicNodeStyleProps | undefined)[]' is not assignable to type 'IntrinsicNodeStyleProps'.
        Index signature for type 'string' is missing in type '(IntrinsicNodeStyleProps | undefined)[]'.ts(2322)
```

File: src/packages/solid/components/ListItem/ListItem.styles.ts [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+ */
+
+import type { NodeStyles } from '@lightningjs/solid';
+import theme from 'theme';
+import type { ComponentStyleConfig, NodeStyleSet, TextStyleSet, Tone } from '../../types/types.js';
+import { getWidthByColumnSpan, makeComponentStyles } from '../../utils/index.js';
+
+export interface ListItemStyles {
+  tone: Tone;
+  Container: NodeStyleSet<{ columnSpan: number; paddingX: number }>;
```
Comment (slhay28) [inline src/packages/solid/components/ListItem/ListItem.styles.ts:25]: I dont think we need columnSpan in here

File: src/packages/solid/components/ListItem/ListItem.tsx [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+      <Text
+        style={[
+          styles.Title.tones[props.tone ?? styles.tone], //
+          styles.Title.base
+        ]}
+      >
+        {props.title}
+      </Text>
+      <Text
+        // y={styles.Description.base.y}
```
Comment (slhay28) [inline src/packages/solid/components/ListItem/ListItem.tsx:76]: remove comment

**Review by slhay28 — APPROVED** [review]

**Pull request merged**